nixos/wireless: add options for better roaming

rnhmjoj opened the pull request:
###### Motivation for this change

See https://github.com/NixOS/nixpkgs/issues/105560#issuecomment-812991613

- Generate a prettier configuration file

- Add an option to automatically launch a scan when the
signal of the current network is low

- Enable 802.11r (fast access point transition) by default for all
protected networks

###### Things done

- [x] Tested 802.11r with two openwrt aps (hostapd says `WPA: FT authentication already completed - do not start 4-way handshake`)
- [x] Tested wpa_supplicant runs
- [x] Tested `userControlled.enable` works
- [x] Built manual
- [x] Tested multiple interfaces
- [x] Tested resuming from sleep with multiple interfaces
- [x] Built and booted iso_minimal    
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rnhmjoj committed:
nixos/wireless: generate pretty configuration

Turns this mess

    ctrl_interface=DIR=/run/wpa_supplicant GROUP=wheel
    update_config=1

    network={
      ssid="cool-network"

      psk="ciao"

    }

    network={
      ssid="fancy-network"

      eap=PEAP
    identity="user@example.com"
    password="secret"

    }

    network={
      ssid="free-network"

      key_mgmt=NONE

    }

    network={
      ssid="raw-network"

      psk=fafafa

    }

into something more human readable:

    network={
      ssid="cool-network"
      psk="ciao"
    }

    network={
      ssid="fancy-network"
      eap=PEAP
      identity="user@example.com"
      password="secret"
    }

    network={
      ssid="free-network"
      key_mgmt=NONE
    }

    network={
      ssid="raw-network"
      psk=fafafa
    }

    ctrl_interface=/run/wpa_supplicant
    ctrl_interface_group=wheel
    update_config=1

rnhmjoj committed:
nixos/wireless: add options for better roaming

- Add an option to automatically launch a scan when the
signal of the current network is low

- Enable 802.11r (fast access point transition) by default for all
protected networks

SuperSandro2000 commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -75,11 +101,48 @@ in {
+              #   /^# [A-Z0-9-]{2,}/{ if(run){printf("\"%s\"\n", $2)} }
+              # ' /run/current-system/sw/share/doc/wpa_supplicant/wpa_supplicant.conf.example
+              type = types.listOf (types.enum [
+                "WPA-PSK"
```
Is WPA2 and WPA3 in here and just named differently?

rnhmjoj replied:
I'm not sure what's you're asking. WPA-Personal is `WPA-PSK` and WPA3-Personal is `SAE`.

bb2020 commented:
Hi. We had a related debate here in #113716. As @Ma27 figured out, it may be problematic to always use `/etc/wpa_supplicant.conf` as `SAVE_CONFIG` command of `wpa_supplicant` overwrites this file.

rnhmjoj commented:
@bb2020 Thank you for bringing this up, I was wondering if this could interfere with `userControlled` somehow.
I'll revert this change but I still would like to link the configuration somewhere: it happens quite often that I need to look at it  and have to:
1. systemctl cat wpa_supplicant
2. open the the ExecStartScript
3. find the `-c /nix/store/....` argument
4. finally open the config file

bb2020 commented:
> 1. systemctl cat wpa_supplicant
> 2. open the the ExecStart script
> 3. find the `-c /nix/store/....` argument
> 4. finally open the config file

If this is needed for non `configuration.nix` SSID entries, then your problem is the same as #113716.
In the meantime, can you try [this](https://github.com/NixOS/nixpkgs/pull/113716#issuecomment-783642967)?

bb2020 commented on nixos/modules/services/networking/wpa_supplicant.nix:
```diff
@@ -47,6 +66,16 @@ in {
 
+      scanOnLowSignal = mkOption {
+        type = types.bool;
+        default = true;
```
I think this should be disabled by default.

rnhmjoj replied:
I'm enabling this by default because it makes roaming "just work". Without this users would probably just assume (as in https://github.com/NixOS/nixpkgs/issues/105560) roaming in wpa_supplicant doesn't work
It can be debated, but I think the benefit of working roaming outweights saving battery percents, though I haven't tried to quantify the impact.

bb2020 replied:
I don't think it impacts battery at all. It will be very confusing for someone who has to move their laptop and `wpa_supplicant` magically changes networks etc.

rnhmjoj replied:
wpa_supplicant (and every other 802.11 client probably) is already doing this: it rescans and connects to the best network available. It's just that, to save battery, it's doing that only if the signal drops to zero.

bb2020 replied:
What about enabling it by default only for minimal ISO? That would solve #105560 while keeping the existing behavior for normal installation.

rnhmjoj replied:
The issue is about using iwd to save space in the minimal install, I don't think enabling roaming only here would help much: when you installing NixOS you're likely not walking around. Proper roaming is useful in you main machine, not the installer.

rnhmjoj committed:
nixos/wireless: add one service per interface

This sets up a different systemd service for each interface. This way
each wpa_supplicant instance waits for his inteface to become ready
using the respective device unit, and that only. The configuration file
is still shared between all instances, though.

This closes a longstanding "fixme" from cbfba81.

rnhmjoj committed:
nixos/wireless: add rnhmjoj as maintainer

rnhmjoj commented:
I've rebased the PR on top of master, fixed a couple issues and rerun all the tests above.
I'm pretty confident with merging this, now.

poscat0x04 commented:
Hi, this pr broke my wireless networking as I am no longer able to authenticate to my wireless network. Do you have any idea what might have caused this?

poscat0x04 commented:
Ok, I've figured out why: my wireless network uses WPA3-SAE for authentication and the `authProtocols` introduced in this PR does not contain WPA3-SAE (by default) and thus broke my wireless networking. I feel like the end-user should be notified when a breaking change like this one is made, especially when it's made to a widely used module like `networking/wireless` @rnhmjoj

rnhmjoj commented:
I haven't written a release note  (yet) because this PR should be backward compatible:
1. By introducing `authProtocols` I've actually enabled more protocols by default than `wpa_supplicant` original default, which is just `key_mgmt=WPA-PSK WPA-EAP`.
2. `WPA3-SAE` doesn't exist[1], if you mean WPA3 Personal, the name is `SAE`, which I've enabled by default in this PR.
3. WPA3 has never been supported (by default) in wpa_supplicant or this module: you would have to enable it by writing something like `wireless.networks.mynetwork.auth="key_mgmt=SAE";` Is this your case?

If you're sure this PR broke your config I need more details to figure out what's wrong. Can you show me your (relevant) `wireless.networks` entry?

[1]: If you want proof try running `set_network 0 key_mgmt WPA3-SAE` in `wpa_cli`.
